### PR TITLE
[fix] add Gemma3 support

### PR DESCRIPTION
**Pull request by kalomaze**
> [!NOTE]
> **Medium Risk**
> Adds a new Gemma-specific autograd path for chunked logprob/entropy and changes meta-device buffer reinit behavior; mistakes here would surface as incorrect logits/loss or failed model loading, but scope is limited to Gemma-configured models.
> 
> **Overview**
> Adds Gemma3 support for meta-device initialization/loading by treating `embed_scale` and a local rotary embedding buffer as reinitializable, and by explicitly recomputing/filling these buffers after `to_empty()`.
> 
> Extends `inject_prime_lm_head` to detect Gemma-style `final_logit_softcapping` and swap in a new Gemma LM head implementation (`lm_head_gemma.py`) that applies softcapping to logits and supports chunked per-token logprob/entropy computation via a custom autograd function, while reusing the same forward patching mechanism.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit bc1b554b06ec4e3b50f6be0699c8295b32f58d94. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/models/layers/lm_head.py [hunk 1]
```diff
@@ -214,16 +239,26 @@ def inject_prime_lm_head(model: nn.Module, chunk_size: int | None = None) -> Non
     )
 
     logger = get_logger()
-    logger.info(f"Injecting Prime LM head with chunk size {chunk_size}")
+    final_logit_softcapping = getattr(model.config, "final_logit_softcapping", None)
```
Comment (S1ro1) [inline src/prime_rl/trainer/models/layers/lm_head.py:227]: Let's maybe keep the old print as is and move the getattr to be close to its usage

File: src/prime_rl/trainer/models/layers/lm_head.py [hunk 2]
```diff
@@ -222,6 +222,15 @@ def inject_prime_lm_head(model: nn.Module, chunk_size: int | None = None) -> Non
...
 
     logger = get_logger()
+
+    # Check for Gemma-style softcapping - dispatch to specialized implementation
+    final_logit_softcapping = getattr(model.config, "final_logit_softcapping", None)
+    if final_logit_softcapping:
+        from prime_rl.trainer.models.layers.lm_head_gemma import inject_gemma_lm_head
+
+        inject_gemma_lm_head(model, chunk_size, final_logit_softcapping)
+        return
```
Comment (S1ro1) [inline src/prime_rl/trainer/models/layers/lm_head.py:232]: I don't really like this early exit, it duplicates a bunch of code between gemma and vanilla inject, wondering if we could maybe do something like:

```python
lm_head_cls = VanillaOutputLinear if chunk_size is None else FusedOutputLinear

lm_head_kwargs = {"in_features": in_features, "out_features": out_features}
if chunk_size:
    lm_head_kwargs["chunk_size"] = chunk_size

if final_logit_softcapping:
    lm_head_cls = GemmaVanillaOutputLinear if chunk_size is None else GemmaFusedOutputLinear
    lm_head_kwargs["softcap"] = softcap

model.lm_head = lm_head_cls(**lm_head_kwargs)
model.lm_head.weight = old_lm_head.weight
del old_lm_head

_patch_model_forward(model)
...
```

**Review by Jackmin801 — APPROVED** [review]
very nice! lgtm!

**Review by S1ro1 — APPROVED** [review]
lgtm

**Review by samsja — APPROVED** [review]

**Comment by SQCU** [comment]
gemma on prime-rl???

**Comment by kalomaze** [comment]
> gemma on prime-rl???

you betcha

File: src/prime_rl/trainer/models/layers/lm_head.py (hunk 2 above)
Comment (kalomaze) [reply src/prime_rl/trainer/models/layers/lm_head.py:232]: i was aiming for loose coupling and independence in a way that wouldnt require lm head to "know" about anything gemma specific
